test/libs/helper-functions.js

```
@@ -0,0 +1,194 @@
+  });
+};
+
+window.testHelper = {
+  unregisterAllRegistrations: function() {
+    return navigator.serviceWorker.getRegistrations()
+      .then(registrations => {
+        if (registrations.length === 0) {
+          return Promise.resolve();
+        }
+
+        var unregisterPromises = [];
+        for (var i = 0; i < registrations.length; i++) {
+          unregisterPromises.push(
+            registrations[i].unregister()
+          );
+        }
+        return Promise.all(unregisterPromises);
+      });
+  },
+
+  clearAllCaches: function() {
+    return window.caches.keys()
+      .then(cacheNames => {
+        if (cacheNames.length === 0) {
+          return Promise.resolve();
+        }
+
+        var cacheTaskPromises = [];
+        for (var i = 0; i < cacheNames.length; i++) {
+          cacheTaskPromises.push(window.caches.delete(cacheNames[i]));
+        }
+        return Promise.all(cacheTaskPromises);
+      });
+  },
+
+  // Waiting for a service worker to install is handy if you only care
+  // about testing events that have occured in the install event
+  installSW: function(swUrl) {
+    return new Promise((resolve, reject) => {
+      var iframe;
+      getIframe()
+      .then(newIframe => {
+        var options = null;
+        if (newIframe) {
+          options = {scope: iframe.contentWindow.location.pathname};
+          iframe = newIframe;
+        }
+
+        return navigator.serviceWorker.register(swUrl, options);
+      })
+      .then(registration => {
+        if (registration.installing === null) {
+          throw new Error(swUrl + ' already installed.');
+        }
+
+        // We unregister all service workers after each test - this should
+        // always trigger an install state change
+        registration.installing.onstatechange = function() {
+          if (this.state !== 'installed') {
+            return;
+          }
+
+          resolve(iframe);
+        };
+      })
+      .catch(err => {
```

My and Paul Lewis have found scenarios where nested callbacks or promises can result in the error getting swallowed and the promise not resolving.